distribution/integration/tests-integration/user-core/src/test/java/org/wso2/carbon/user/core/ldap/ReadOnlyLDAPUserStoreManagerTest.java

```diff
@@ -0,0 +1,285 @@
+ * Uses embedded LDAP to check the functionality of the class.
+ *
+ */
+public class ReadOnlyLDAPUserStoreManagerTest {
+
+    private static final Log log = LogFactory.getLog(ReadOnlyLDAPUserStoreManagerTest.class);
+
+    private static final String MOBILE_CLAIM = "www.wso2.org/mobile";
+    private static final String NAME_CLAIM = "www.wso2.org/name";
+    private static final String SNAME_CLAIM = "www.wso2.org/surname";
+    private static final String U_NAME_CLAIM = "www.wso2.org/uName";
+    private static final String MOBILE_ATTR = "mobile";
+    private static final String CN_ATTR = "cn";
+    private static final String SNAME_ATTR = "sn";
+    private static final String U_NAME_ATTR = "uName";
+    private static final String USER_NAME_1 = "LName1";
+    private static final String USER_NAME_2 = "LName2";
+    private static final String USER_1_U_NAME = "U_Name_1";
+    private static final String USER_UPDATED_LAST_NAME = "LNameX";
+    private static final String SEARCH_BASE = "ou=Users,dc=WSO2,dc=ORG";
+    private static final int LDAP_SERVER_PORT = 12389;
+
+    private JdbcDataSource dataSource;
+    private LDAPConnectionContext ldapConnectionContext;
+    private ClaimManager claimManager;
+    private Map<String, String> claimMap = new HashMap<String, String>();
+    private TestDirectoryServer testDirectoryServer;
+
+    @BeforeSuite
+    public void setUp() throws Exception {
+        System.setProperty("carbon.home", ReadOnlyLDAPUserStoreManagerTest.class.getResource("/").getFile());
+        MBeanServer server = MBeanServerFactory.newMBeanServer();
+        dataSource = createDatasource();
+        ldapConnectionContext = new LDAPConnectionContext(getRealmConfiguration());
+        claimManager = Mockito.mock(ClaimManager.class);
+
+        claimMap.put(MOBILE_CLAIM, MOBILE_ATTR);
+        claimMap.put(NAME_CLAIM, CN_ATTR);
+        claimMap.put(SNAME_CLAIM, SNAME_ATTR);
+        claimMap.put(U_NAME_CLAIM, U_NAME_ATTR);
+
+        Mockito.doAnswer(new Answer() {
+
+            public Object answer(InvocationOnMock invocationOnMock) throws Throwable {
+                return claimMap.get(invocationOnMock.getArguments()[0]);
+            }
+        }).when(claimManager).getAttributeName(Matchers.anyString());
+        Mockito.doAnswer(new Answer() {
+
+            public Object answer(InvocationOnMock invocationOnMock) throws Throwable {
+                ClaimMapping result = new ClaimMapping();
+                result.setMappedAttribute(claimMap.get(invocationOnMock.getArguments()[0]));
+                return result;
+            }
+        }).when(claimManager).getClaimMapping(Matchers.anyString());
+
+        testDirectoryServer = new TestDirectoryServer();
+        testDirectoryServer.startLdapServer(LDAP_SERVER_PORT);
+
+        clearTestUsers();
+    }
+
+    @AfterSuite
+    public void tearDown() throws UserStoreException {
+        clearTestUsers();
+        testDirectoryServer.stopLdapService();
```

Comment:
What would happen #clearTestUsers threw an exception? Will the LDAP port kept open?
May be the #stopLdapService need to go to a finally block.

Reply:
+1